changelog for 1.0 (#2440)
jupyterhub/jupyterhub: docs/source/changelog.md

## [Unreleased]

## 1.0

### [1.0.0] 2018-03-XX

JupyterHub 1.0 is a major milestone for JupyterHub.
Huge thanks to the many people who have contributed to this release,

> whether it was through discussion, testing, documentation, or code.

Reviewer: ```suggestion
whether it was through discussion, testing, documentation, or development.
```

#### Major new features

- Support TLS encryption and authentication of all internal communication.
  Spawners must implement `.move_certs` method to make certificates available
  to the notebook server if it is not local to the Hub.

> - There is now full UI support for managing named servers (TODO: screenshot).

Reviewer: - [ ] Add screenshot

>   With named servers, each jupyterhub user may have more than one named server.

Reviewer: ```suggestion
  With named servers, each jupyterhub user may have access to more than one named server. For example, a professor may access a server named `research` and another named `teaching`.
```

- Authenticators can now expire and refresh authentication data by implementing
  `Authenticator.refresh_user(user)`.
  This allows things like OAuth data and access tokens to be refreshed.

>   By setting `Authenticator.refresh_pre_spawn = True`,

Reviewer: ```suggestion
  When used together with `Authenticator.refresh_pre_spawn = True`,
```

  auth refresh can be forced prior to Spawn,
  allowing the Authenticator to *require* that authentication data is fresh
  immediately before the user's server is launched.

```eval_rst
.. seealso::

  - :meth:`.Authenticator.refresh_user`
  - :meth:`.Spawner.create_certs`
  - :meth:`.Spawner.move_certs`
```

#### New features

> - allow spawners, authenticators, and proxies to register themselves via 'entry points', enabling more convenient configuration such as:

Reviewer: ```suggestion
- allow custom spawners, authenticators, and proxies to register themselves via 'entry points', enabling more convenient configuration such as:
```

  ```python
  c.JupyterHub.authenticator_class = 'github'
  c.JupyterHub.spawner_class = 'docker'
  c.JupyterHub.proxy_class = 'traefik_etcd'
  ```
- Spawners are passed the tornado Handler object that requested their spawn (as `self.handler`),
  so they can do things like make decisions based on query arguments in the request.
- SimpleSpawner and DummyAuthenticator, which are useful for testing, have been merged into JupyterHub itself:

>   ```python

Reviewer: ```suggestion
  ```python
  # For testing purposes only. Should not be used in production.
```

  c.JupyterHub.authenticator_class = 'dummy'
  c.JupyterHub.spawner_class = 'simple'
  ```

  These classes are **not** appropriate for production use. Only testing.
- Add health check endpoint at `/hub/health`

> - Several prometheus metrics have been added (thanks to Outreachy applicants!)

Reviewer: ```suggestion
- Several prometheus metrics have been added (thanks to [Outreachy](https://www.outreachy.org/) applicants!)
```

- A new API for registering user activity.

>   To prepare for [alternate proxy implementations](https://github.com/jupyterhub/traefik-proxy),

Reviewer: ```suggestion
  To prepare for the addition of [alternate proxy implementations](https://github.com/jupyterhub/traefik-proxy),
```

  responsibility for tracking activity is taken away from the proxy
  and moved to the notebook server (which already has activity tracking features).

>   Activity is pushed to the Hub from user servers instead of polled from the

Reviewer: ```suggestion
  Activity is now tracked by pushing it to the Hub from user servers instead of polling the
```

  proxy API.
- Dynamic `options_form` callables may now return an empty string
  which will result in no options form being rendered.

#### Changes

- Prometheus metrics page is now authenticated. Any authenticated user may see the prometheus metrics. To disable prometheus authentication, set `JupyterHub.authenticate_prometheus = False`
- Visits to `/user/:name` no longer trigger an implicit launch of the user's server.
  Instead, a page is shown indicating that the server is not running
  with a link to request the spawn.

> - API requests to `/user/:name` for a not-running server will have status 503 instead of 404.

Reviewer: not-running or stopped?

Author: not-running, I think. It could be stopped or never started.

#### Fixed

- Various fixes to improve Windows compatibility
  (default Authenticator and Spawner still do not support Windows, but other Spawners may)

#### Development changes

> There have been several changes to development process that shouldn't

Reviewer: ```suggestion
There have been several changes to the development process that shouldn't
```

> generally affect users of JupyterHub, but may affect folks contributing to it.

Reviewer: ```suggestion
generally affect users of JupyterHub, but may affect contributors.
```
